File: templates/Uwp/_composition/_shared/Page.Ink.AddTransformService/Services/Ink/InkTransformService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Windows.Foundation;
using Windows.UI;
using Windows.UI.Input.Inking;
using Windows.UI.Input.Inking.Analysis;
using Windows.UI.Xaml;
using Windows.UI.Xaml.Controls;
using Windows.UI.Xaml.Media;
using Windows.UI.Xaml.Shapes;

namespace Param_ItemNamespace.Services.Ink
{
    public class InkTransformService
    {
        private readonly InkAnalyzer _inkAnalyzer;
        private readonly Canvas _drawingCanvas;
        private readonly InkStrokesService _strokeService;

        public InkTransformService(Canvas drawingCanvas, InkStrokesService strokeService)
        {
            _drawingCanvas = drawingCanvas;
            _strokeService = strokeService;
            _inkAnalyzer = new InkAnalyzer();
        }

        public async Task<InkTransformResult> TransformTextAndShapesAsync()
        {
            var result = new InkTransformResult(_drawingCanvas);
            var inkStrokes = GetStrokesToConvert();

            if (inkStrokes.Any())
            {
                _inkAnalyzer.ClearDataForAllStrokes();
                _inkAnalyzer.AddDataForStrokes(inkStrokes);
                var inkAnalysisResults = await _inkAnalyzer.AnalyzeAsync();

                if (inkAnalysisResults.Status == InkAnalysisStatus.Updated)
                {
                    var words = AnalyzeWords();
                    var shapes = AnalyzeShapes();

                    // Generate result
                    result.Strokes.AddRange(inkStrokes);
                    result.TextAndShapes.AddRange(words);
                    result.TextAndShapes.AddRange(shapes);
                }
            }

            return result;
        }

        public void ClearTextAndShapes()
        {
            _drawingCanvas.Children.Clear();
        }

        private IEnumerable<UIElement> AnalyzeWords()
        {
            var inkwordNodes = _inkAnalyzer.AnalysisRoot.FindNodes(InkAnalysisNodeKind.InkWord);

            foreach (InkAnalysisInkWord node in inkwordNodes)
            {
                var textblock = DrawText(node.RecognizedText, node.BoundingRect);

                var strokesIds = node.GetStrokeIds();
                _strokeService.RemoveStrokesByIds(strokesIds);
                _inkAnalyzer.RemoveDataForStrokes(strokesIds);

                yield return textblock;
            }
        }

        private IEnumerable<UIElement> AnalyzeShapes()
        {
            var inkdrawingNodes = _inkAnalyzer.AnalysisRoot.FindNodes(InkAnalysisNodeKind.InkDrawing);
            foreach (InkAnalysisInkDrawing node in inkdrawingNodes)
            {
                var strokesIds = node.GetStrokeIds();

                if (node.DrawingKind == InkAnalysisDrawingKind.Drawing)
                {
                    // Catch and process unsupported shapes (lines and so on) here.
                }
                else
                {
                    if (node.DrawingKind == InkAnalysisDrawingKind.Circle || node.DrawingKind == InkAnalysisDrawingKind.Ellipse)
                    {
                        yield return DrawEllipse(node);
                    }
                    else
                    {
                        yield return DrawPolygon(node);
                    }

                    _strokeService.RemoveStrokesByIds(strokesIds);
                }

                _inkAnalyzer.RemoveDataForStrokes(strokesIds);
            }
        }

        private UIElement DrawText(string recognizedText, Rect boundingRect)
        {
            TextBlock text = new TextBlock();
            Canvas.SetTop(text, boundingRect.Top);
            Canvas.SetLeft(text, boundingRect.Left);

            text.Text = recognizedText;
            text.FontSize = boundingRect.Height;
            text.Foreground = new SolidColorBrush(Colors.Black);

            _drawingCanvas.Children.Add(text);
            return text;
        }

        private UIElement DrawEllipse(InkAnalysisInkDrawing shape)
        {
            var points = shape.Points;
            Ellipse ellipse = new Ellipse();
            ellipse.Width = Math.Sqrt(((points[0].X - points[2].X) * (points[0].X - points[2].X)) +
                 ((points[0].Y - points[2].Y) * (points[0].Y - points[2].Y)));
            ellipse.Height = Math.Sqrt(((points[1].X - points[3].X) * (points[1].X - points[3].X)) +
                 ((points[1].Y - points[3].Y) * (points[1].Y - points[3].Y)));

            var rotAngle = Math.Atan2(points[2].Y - points[0].Y, points[2].X - points[0].X);
            RotateTransform rotateTransform = new RotateTransform();
            rotateTransform.Angle = rotAngle * 180 / Math.PI;
            rotateTransform.CenterX = ellipse.Width / 2.0;
            rotateTransform.CenterY = ellipse.Height / 2.0;

            TranslateTransform translateTransform = new TranslateTransform();
            translateTransform.X = shape.Center.X - (ellipse.Width / 2.0);
            translateTransform.Y = shape.Center.Y - (ellipse.Height / 2.0);

            TransformGroup transformGroup = new TransformGroup();
            transformGroup.Children.Add(rotateTransform);
            transformGroup.Children.Add(translateTransform);
            ellipse.RenderTransform = transformGroup;

            var brush = new SolidColorBrush(ColorHelper.FromArgb(255, 0, 0, 255));
            ellipse.Stroke = brush;
            ellipse.StrokeThickness = 2;
            _drawingCanvas.Children.Add(ellipse);

            return ellipse;
        }

        private UIElement DrawPolygon(InkAnalysisInkDrawing shape)
        {
            var points = shape.Points;
            Polygon polygon = new Polygon();

            foreach (var point in points)
            {
                polygon.Points.Add(point);
            }

            var brush = new SolidColorBrush(Windows.UI.ColorHelper.FromArgb(255, 0, 0, 255));
            polygon.Stroke = brush;
            polygon.StrokeThickness = 2;
            _drawingCanvas.Children.Add(polygon);

            return polygon;
        }

        private IEnumerable<InkStroke> GetStrokesToConvert()
        {
            var selectedStrokes = _strokeService.GetSelectedStrokes();

            if (selectedStrokes.Any())
            {
                return selectedStrokes;
            }

            return _strokeService.GetStrokes();
        }
    }
}

File: code/TemplateStudioForUWP/Templates/Uwp/_comp/_shared/Page.Ink.AddTransformService._VB/Services/Ink/InkTransformService.vb
Imports Windows.UI
Imports Windows.UI.Input.Inking
Imports Windows.UI.Input.Inking.Analysis
Imports Windows.UI.Xaml.Shapes

Namespace Services.Ink
    Public Class InkTransformService
        Private ReadOnly _inkAnalyzer As InkAnalyzer
        Private ReadOnly _drawingCanvas As Canvas
        Private ReadOnly _strokeService As InkStrokesService

        Public Sub New(drawingCanvas As Canvas, strokeService As InkStrokesService)
            _drawingCanvas = drawingCanvas
            _strokeService = strokeService
            _inkAnalyzer = New InkAnalyzer()
        End Sub

        Public Async Function TransformTextAndShapesAsync() As Task(Of InkTransformResult)
            Dim result = New InkTransformResult(_drawingCanvas)
            Dim inkStrokes = GetStrokesToConvert()

            If inkStrokes.Any() Then
                _inkAnalyzer.ClearDataForAllStrokes()
                _inkAnalyzer.AddDataForStrokes(inkStrokes)
                Dim inkAnalysisResults = Await _inkAnalyzer.AnalyzeAsync()

                If inkAnalysisResults.Status = InkAnalysisStatus.Updated Then
                    Dim words = AnalyzeWords()
                    Dim shapes = AnalyzeShapes()

                    ' Generate result
                    result.Strokes.AddRange(inkStrokes)
                    result.TextAndShapes.AddRange(words)
                    result.TextAndShapes.AddRange(shapes)
                End If
            End If

            Return result
        End Function

        Public Function GetTextAndShapes() As IEnumerable(Of UIElement)
            Return _drawingCanvas.Children
        End Function

        Public Sub AddUIElement(element As UIElement)
            _drawingCanvas.Children.Add(element)
        End Sub

        Public Sub RemoveUIElement(element As UIElement)
            If _drawingCanvas.Children.Contains(element) Then
                _drawingCanvas.Children.Remove(element)
            End If
        End Sub

        Public Function HasTextAndShapes as Boolean
            Return _drawingCanvas.Children.Any()
        End Function

        Public Sub ClearTextAndShapes()
            _drawingCanvas.Children.Clear()
        End Sub

        Private Iterator Function AnalyzeWords() As IEnumerable(Of UIElement)
            Dim inkwordNodes = _inkAnalyzer.AnalysisRoot.FindNodes(InkAnalysisNodeKind.InkWord)

            For Each node As InkAnalysisInkWord In inkwordNodes
                Dim textblock = DrawText(node.RecognizedText, node.BoundingRect)
                Dim strokesIds = node.GetStrokeIds()
                _strokeService.RemoveStrokesByIds(strokesIds)
                _inkAnalyzer.RemoveDataForStrokes(strokesIds)
                Yield textblock
            Next
        End Function

        Private Iterator Function AnalyzeShapes() As IEnumerable(Of UIElement)
            Dim inkdrawingNodes = _inkAnalyzer.AnalysisRoot.FindNodes(InkAnalysisNodeKind.InkDrawing)

            For Each node As InkAnalysisInkDrawing In inkdrawingNodes
                Dim strokesIds = node.GetStrokeIds()

                If node.DrawingKind = InkAnalysisDrawingKind.Drawing Then
                    ' Catch and process unsupported shapes (lines and so on) here.
                Else

                    If node.DrawingKind = InkAnalysisDrawingKind.Circle OrElse node.DrawingKind = InkAnalysisDrawingKind.Ellipse Then
                        Yield DrawEllipse(node)
                    Else
                        Yield DrawPolygon(node)
                    End If

                    _strokeService.RemoveStrokesByIds(strokesIds)
                End If

                _inkAnalyzer.RemoveDataForStrokes(strokesIds)
            Next
        End Function

        Private Function DrawText(recognizedText As String, boundingRect As Rect) As UIElement
            Dim text As TextBlock = New TextBlock()
            Canvas.SetTop(text, boundingRect.Top)
            Canvas.SetLeft(text, boundingRect.Left)
            text.Text = recognizedText
            text.FontSize = boundingRect.Height
            text.Foreground = New SolidColorBrush(Colors.Black)
            _drawingCanvas.Children.Add(text)
            Return text
        End Function

        Private Function DrawEllipse(shape As InkAnalysisInkDrawing) As UIElement
            Dim points = shape.Points
            Dim ellipse As Ellipse = New Ellipse()
            ellipse.Width = Math.Sqrt(((points(0).X - points(2).X) * (points(0).X - points(2).X)) + ((points(0).Y - points(2).Y) * (points(0).Y - points(2).Y)))
            ellipse.Height = Math.Sqrt(((points(1).X - points(3).X) * (points(1).X - points(3).X)) + ((points(1).Y - points(3).Y) * (points(1).Y - points(3).Y)))
            Dim rotAngle = Math.Atan2(points(2).Y - points(0).Y, points(2).X - points(0).X)
            Dim rotateTransform As RotateTransform = New RotateTransform()
            rotateTransform.Angle = rotAngle * 180 / Math.PI
            rotateTransform.CenterX = ellipse.Width / 2.0
            rotateTransform.CenterY = ellipse.Height / 2.0
            Dim translateTransform As TranslateTransform = New TranslateTransform()
            translateTransform.X = shape.Center.X - (ellipse.Width / 2.0)
            translateTransform.Y = shape.Center.Y - (ellipse.Height / 2.0)
            Dim transformGroup As TransformGroup = New TransformGroup()
            transformGroup.Children.Add(rotateTransform)
            transformGroup.Children.Add(translateTransform)
            ellipse.RenderTransform = transformGroup
            Dim brush = New SolidColorBrush(ColorHelper.FromArgb(255, 0, 0, 255))
            ellipse.Stroke = brush
            ellipse.StrokeThickness = 2
            _drawingCanvas.Children.Add(ellipse)
            Return ellipse
        End Function

        Private Function DrawPolygon(shape As InkAnalysisInkDrawing) As UIElement
            Dim points = shape.Points
            Dim polygon As Polygon = New Polygon()

            For Each point In points
                polygon.Points.Add(point)
            Next

            Dim brush = New SolidColorBrush(ColorHelper.FromArgb(255, 0, 0, 255))
            polygon.Stroke = brush
            polygon.StrokeThickness = 2
            _drawingCanvas.Children.Add(polygon)
            Return polygon
        End Function

        Private Function GetStrokesToConvert() As IEnumerable(Of InkStroke)
            Dim selectedStrokes = _strokeService.GetSelectedStrokes()

            If selectedStrokes.Any() Then
                Return selectedStrokes
            End If

            Return _strokeService.GetStrokes()
        End Function
    End Class
End Namespace
